consensus/hotstuff/timeoutaggregator/timeout_aggregator_test.go

```diff
@@ -0,0 +1,141 @@
+package timeoutaggregator
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/atomic"
+
+	"github.com/onflow/flow-go/consensus/hotstuff/helper"
+	"github.com/onflow/flow-go/consensus/hotstuff/mocks"
+	"github.com/onflow/flow-go/consensus/hotstuff/model"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+func TestTimeoutAggregator(t *testing.T) {
+	suite.Run(t, new(TimeoutAggregatorTestSuite))
+}
+
+// TimeoutAggregatorTestSuite is a test suite for isolated testing of TimeoutAggregator.
+// Contains mocked state which is used to verify correct behavior of TimeoutAggregator.
+// Automatically starts and stops module.Startable in SetupTest and TearDownTest respectively.
+type TimeoutAggregatorTestSuite struct {
+	suite.Suite
+
+	lowestRetainedView uint64
+	highestKnownView   uint64
+	aggregator         *TimeoutAggregator
+	collectors         *mocks.TimeoutCollectors
+	committee          *mocks.Replicas
+	consumer           *mocks.Consumer
+	stopAggregator     context.CancelFunc
+}
+
+func (s *TimeoutAggregatorTestSuite) SetupTest() {
+	var err error
+	s.collectors = mocks.NewTimeoutCollectors(s.T())
+	s.consumer = mocks.NewConsumer(s.T())
+	s.committee = mocks.NewReplicas(s.T())
+
+	s.lowestRetainedView = 100
+
+	s.committee.On("LeaderForView", mock.Anything).Return(flow.Identifier{}, nil).Maybe()
+
+	s.aggregator, err = NewTimeoutAggregator(unittest.Logger(), s.consumer, s.lowestRetainedView, s.committee, s.collectors)
+	require.NoError(s.T(), err)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	signalerCtx, _ := irrecoverable.WithSignaler(ctx)
+	s.stopAggregator = cancel
+	s.aggregator.Start(signalerCtx)
+	unittest.RequireCloseBefore(s.T(), s.aggregator.Ready(), 100*time.Millisecond, "should close before timeout")
+}
+
+func (s *TimeoutAggregatorTestSuite) TearDownTest() {
+	s.stopAggregator()
+	unittest.RequireCloseBefore(s.T(), s.aggregator.Done(), time.Second, "should close before timeout")
+}
+
+// TestAddTimeout_HappyPath tests a happy path when multiple threads are adding timeouts for processing
+// Eventually every timeout has to be processed by TimeoutCollector
+func (s *TimeoutAggregatorTestSuite) TestAddTimeout_HappyPath() {
+	timeoutsCount := 20
+	collector := mocks.NewTimeoutCollector(s.T())
+	callCount := atomic.NewUint64(0)
+	collector.On("AddTimeout", mock.Anything).Run(func(mock.Arguments) {
+		callCount.Add(1)
+	}).Return(nil).Times(timeoutsCount)
+	s.collectors.On("GetOrCreateCollector", s.lowestRetainedView).Return(collector, true, nil).Times(timeoutsCount)
+
+	var wg, startWg sync.WaitGroup
+	wg.Add(timeoutsCount)
+	startWg.Add(1)
```

Comment:
`wg.Wait` doesn't give guarantee that calls indeed were made since `AddTimeout` is async so we definitely need to use `callCount`. I've forgot to wait on `startWg`, the idea was to delay start of processing until all goroutines are spawned. Update it.

Reply:
Yeah, I think we can remove it.